pkg/provisioner/crdprovisioner.go

```diff
@@ -250,58 +290,68 @@ func (c *CrdProvisioner) PublishVolume(
 
 	attachmentName := azureutils.GetAzVolumeAttachmentName(volumeName, nodeID)
 
-	azVolumeAttachmentInstance, err := azVA.Get(ctx, attachmentName, metav1.GetOptions{})
+	if volumeContext == nil {
+		volumeContext = map[string]string{}
+	}
+	azVolumeAttachmentInstance, err := azVA.Get(ctx, strings.ToLower(attachmentName), metav1.GetOptions{})
 	if err == nil {
-		if azVolumeAttachmentInstance.Status != nil && azVolumeAttachmentInstance.Status.Error != nil {
-			azVolumeAttachmentInstance.Status = nil
-			azVolumeAttachmentUpdated, err := azVA.UpdateStatus(ctx, azVolumeAttachmentInstance, metav1.UpdateOptions{})
-			if err != nil || azVolumeAttachmentUpdated == nil || azVolumeAttachmentUpdated.Status != nil {
-				klog.Errorf("Failed to update azVolumeAttachment resource for volume name (%s), error: %v", volumeName, err)
-				return nil, err
-			}
-		} else {
-
-			// If there exists an attachment, we are trying to update
-			// the AzVolumeAttachment role from Replica to Primary
-			azVolumeAttachmentInstance.Spec.RequestedRole = v1alpha1.PrimaryRole
-			_, err := azVA.Update(ctx, azVolumeAttachmentInstance, metav1.UpdateOptions{})
-			if err != nil {
+		updated := azVolumeAttachmentInstance.DeepCopy()
+		updated.Spec.RequestedRole = v1alpha1.PrimaryRole
+		if updated.Status != nil && updated.Status.Error != nil {
+			updated.Status = nil
+			updated, err = azVA.UpdateStatus(ctx, updated, metav1.UpdateOptions{})
+			if err != nil || updated == nil || updated.Status != nil {
+				klog.Errorf("Failed to update azVolumeAttachment resource status for volume name (%s), error: %v", volumeName, err)
 				return nil, err
 			}
-			return azVolumeAttachmentInstance.Status.PublishContext, nil
 		}
-	}
 
-	azVolumeAttachment := &v1alpha1.AzVolumeAttachment{
-		ObjectMeta: metav1.ObjectMeta{
-			Name: attachmentName,
-			Labels: map[string]string{
-				"node-name":   nodeID,
-				"volume-name": volumeName,
+		// If there exists an attachment, we are trying to update
+		// the AzVolumeAttachment role from Replica to Primary
+		updated.Spec.RequestedRole = v1alpha1.PrimaryRole
+		_, err := azVA.Update(ctx, updated, metav1.UpdateOptions{})
+		if err != nil {
+			return nil, err
+		}
+	} else if !errors.IsNotFound(err) {
```

Comment:
@mukhoakash This didn't seem to work when I was testing. When the driver couldn't find the azVolumeAttachment object, it failed with errorCode Unknown instead of NotFound. Have you also encountered this?

Reply:
> > > @nearora-msft One quick question which I wanted to ask here is, since the azVolume is fetched using the volumeName and if on finding that azVolume, we are updating the status of that resource without checking if any of the other parameters in the request has changed or not with what it is present in the spec.
> > > Do you think we need to handle this? @sunpa93 Was wondering the same about the change in `PublishVolume`.
> > 
> > 
> > I believe the only time we are updating the `Status` field (setting it `nil`) is when we have a non-nil `Status.Error`. If `Status.Error != nil`, the underlying volume creation or volume attachment has failed for whatever reason in previous attempt. And with the new error-handling logic in `AzVolume` and `AzVolumeAttachment` controller, the controllers will no longer retry failed attachment or volume creation on its own until the `crdprovisioner` resets the `Status` field to `nil`. So, in order to trigger the controllers' retry, we need to set the `Status` field to `nil`.
> > So, I don't think there is any field in CRI's spec that should affect this behavior. @mukhoakash, was there any field in spec that you specifically had in mind that needs handling?
> 
> Yeah I agree with @sunpa93 here. But yeah @mukhoakash I understand that there's an underlying assumption here that the since the call is being retried by the sidecars, the spec would be the same.

I discussed this with @sunpa93  regarding the `PublishVolume`'s behavior and we agreed that for the sake of safety, we will update all the fields in the spec when the CR exists in a failed state prior to the current request. 
@nearora-msft I think it wouldn't harm to address the same for 'CreateVolume` as well?

Reply:
I brought back some of the overwritten code due to merge conflicts. I didnt encounter this myself. @sunpa93  Did you encounter this?

Reply:
I haven't encountered this but if it does happen occasionally it could be problematic.

Reply:
I am planning to test this scenario on this PR again to confirm when it happens and how can we avoid it.

Reply:
I tested it again. Looks like errors.IsNotFound(err) works and returns appropriate result. I  tried to to understand why I was getting UNKNOWN and realized that I was trying to test it by checking the error code i.e. by checking if _status.Code(err) == codes.NotFound_ and this was failing because the _status.Code(err) = UNKNOWN_ in this scenario. 
Sorry for the confusion.